client-js/app/elements/labrad-grapher.ts

```diff
@@ -134,41 +185,61 @@ export class LabradGrapher extends polymer.Base {
         trashed: false
       });
     }
+
     for (let dir of dirs) {
-      if (isIncluded(dir)) {
-        items.push({
-          id: 'dir:' + dir.name,
-          name: dir.name,
-          url: dir.url,
-          isDir: true,
-          isDataset: false,
-          isParent: false,
-          starred: isTagged(dir, 'star'),
-          trashed: isTagged(dir, 'trash')
-        });
-      }
+      this.newDir(dir);
     }
+
     for (let dataset of datasets) {
-      if (isIncluded(dataset)) {
-        items.push({
-          id: 'dataset:' + dataset.name,
-          name: dataset.name,
-          url: dataset.url,
-          isDir: false,
-          isDataset: true,
-          isParent: false,
-          starred: isTagged(dataset, 'star'),
-          trashed: isTagged(dataset, 'trash')
-        });
-      }
+      this.newDataset(dataset);
+    }
+  }
+
+  private addItemToList_(item: ListItem): void {
+    this.push('listItems_', item);
+    if (this.filterListItems_([item]).length == 1) {
+      this.push('listItems', item);
     }
-    return items;
+  }
+
+  private filterListItems_(items: Array<ListItem>) {
+    if (this.showTrash) {
+      return items.filter((x) => {
+        return (x.trashed || x.isParent);
+      });
+    }
+
+    if (this.showStars) {
+      return items.filter((x) => {
+        return (x.starred && !x.trashed) || x.isParent;
+      });
+    }
+
+    return items.filter((x) => {
+      return !x.trashed || x.isParent;
+    });
+  }
+
+  private applyListFilter_(): void {
+    // Will cause `iron-list` to jump to the top.
+    this.set('listItems', this.filterListItems_(this.listItems_));
+  }
+
+  private isTagged_(item: {tags: Array<string>}, tag: string) {
+    for (let t of item.tags) {
+      if (t === tag) return true;
+    }
+    return false;
   }
 
   @computed()
-  selectedDataset(selectedId: string): string {
-    if (selectedId.startsWith('dataset:')) {
-      var name = selectedId.substring(8);
+  selectedDataset(selected: {id: string}): string {
+    if (!selected) {
+      return "";
+    }
+
+    if (selected.id.startsWith('dataset:')) {
```

Comment:
Separate PR would be fine; best to keep these small and focused. The state of this code is such that for almost every line that gets touched in a PR, one can see a few other things that need fixing... Glad you're helping fix things up :-)